javalib/src/main/scala/java/lang/Integer.scala

```diff
@@ -483,9 +482,11 @@ object Integer {
 
       if (digit == -1) throw new NumberFormatException(s)
 
-      if (compareUnsigned(result, max) == 1) throw new NumberFormatException(s)
+      val next = result * radix + digit
 
-      result = result * radix + digit
+      if (compareUnsigned(next, result) < 0) throw new NumberFormatException(s)
```

Comment:
You cannot use comparison after the fact to check for overflow of *multiplication*. That's because multiplication overflow does not "wrap around" in any meaningful way. It just overflows and bits are lost.

For example, in 8 bits, consider `80 * 8`. 80 is `01010000` in binary. If `result == 80`, `radix == 8` and `digit == 0` at this point, then `next = result * radix + digit == 80 * 8 + 0 == 80 << 3 == 01010000b << 3 == 10000000b == 128`. You have `next` is unsigned-greater than `result`, but you had an overflow.

The previous check with `max` that you had was the good way to test for overflow of the *multiplication*. My addition is the good way to test for overflow of the *addition*. You need both.